x-pack/filebeat/input/netflow/integration_test.go

```diff
@@ -0,0 +1,388 @@
+					"queue.mem.events":           12800,
+					"queue.mem.flush.min_events": 1600,
+					"queue.mem.flush.timeout":    5,
+					"compression_level":          1,
+					"connection_idle_timeout":    15,
+				}),
+			},
+		},
+		{
+			Id:             "input-unit-1",
+			Type:           proto.UnitType_INPUT,
+			ConfigStateIdx: 0,
+			State:          proto.State_HEALTHY,
+			LogLevel:       proto.UnitLogLevel_INFO,
+			Config: &proto.UnitExpectedConfig{
+				Id:   "netflow-netflow-1e8b33de-d54a-45cd-90da-23ed71c482e5",
+				Type: "netflow",
+				Name: "netflow-1",
+				Source: integration.RequireNewStruct(t, map[string]interface{}{
+					"use_output": "default",
+					"revision":   0,
+				}),
+				DataStream: &proto.DataStream{
+					Namespace: "default",
+				},
+				Meta: &proto.Meta{
+					Package: &proto.Package{
+						Name:    "netflow",
+						Version: "1.9.0",
+					},
+				},
+				Streams: []*proto.Stream{
+					{
+						Id: "netflow-netflow.netflow-1e8b33de-d54a-45cd-90da-23ed71c482e2",
+						DataStream: &proto.DataStream{
+							Dataset: "netflow.log",
+						},
+						Source: integration.RequireNewStruct(t, map[string]interface{}{
+							"id":                    "netflow_integration_test",
+							"host":                  "localhost:6006",
+							"expiration_timeout":    "30m",
+							"queue_size":            2 * 4 * 1600,
+							"detect_sequence_reset": true,
+							"max_message_size":      "10KiB",
+						}),
+					},
+				},
+			},
+		},
+	}
+
+	healthyChan := make(chan struct{})
+	server := &mock.StubServerV2{
+		CheckinV2Impl: func(observed *proto.CheckinObserved) *proto.CheckinExpected {
+
+			if healthyChan != nil {
+				unitState, payload := extractStateAndPayload(observed, "input-unit-1")
+				if unitState == proto.State_HEALTHY {
+					if payload.streamStatusEquals("netflow-netflow.netflow-1e8b33de-d54a-45cd-90da-23ed71c482e2", map[string]interface{}{
+						"status": "HEALTHY",
+						"error":  "",
+					}) {
+						close(healthyChan)
+						healthyChan = nil
+					}
+				}
+			}
+
+			return &proto.CheckinExpected{
+				Units: allStreams,
+			}
+		},
+		ActionImpl: func(response *proto.ActionResponse) error {
+			return nil
+		},
+	}
+	if err := server.Start(); err != nil {
+		t.Fatalf("failed to start StubServerV2 server: %v", err)
+	}
+	defer server.Stop()
+
+	// It's necessary to change os.Args so filebeat.Filebeat() can read the
+	// appropriate args at beat.Execute().
+	initialOSArgs := os.Args
+	os.Args = []string{
+		"filebeat",
+		"-E", fmt.Sprintf(`management.insecure_grpc_url_for_testing="localhost:%d"`, server.Port),
+		"-E", "management.enabled=true",
+		"-E", "management.restart_on_output_change=true",
+		"-E", "logging.level=info",
+	}
+	defer func() {
+		os.Args = initialOSArgs
+	}()
+
+	beatCmd := filebeat.Filebeat()
+	beatRunErr := make(chan error)
+	go func() {
+		defer close(beatRunErr)
+		beatRunErr <- beatCmd.Execute()
+	}()
+
+	select {
+	case <-healthyChan:
```

Comment:
The ```break``` is redundant, can be removed

I think there might be a chance of the race.
The "select" and the server (callbacks) are on separate go routines.

I can't reproduce this easily, but since there is no synchronization I think depending on the runtime scheduler timing you could have two situations here:
1. healthyChan is closed: which breaks out of select 
2. healthyChan is nil: which will block forever, waiting for other channels (worst case it's the test flakiness, timing out 10 secs timer chan).

The simple fix could be just to use context.Context instead of the healthyChan, it's goroutine-safe.

Reply:
> The break is redundant, can be removed

yes it is but I like it as it gives clear intentions to a new comer and it doesn't alternate the behaviour of the code?

>I can't reproduce this easily, but since there is no synchronization I think depending on the runtime scheduler timing you could have two situations here:

>healthyChan is closed: which breaks out of select
healthyChan is nil: which will block forever, waiting for other channels (worst case it's the test flakiness, timing out 10 secs timer chan).
> The simple fix could be just to use context.Context instead of the healthyChan, it's goroutine-safe.

you can't because this indeed async server callback is called way after the beatCmd.Execute() and right after the goroutine that invokes that I am at select. But I see your point and I don't like even tiny race windows so I am gonna address it

Reply:
race condition eliminated with sync.Once; thanks for the proposal @aleksmaus